server/src/main/java/org/jboss/as/server/ServerService.java

```diff
@@ -312,9 +312,14 @@ protected void boot(final BootContext context) throws ConfigurationPersistenceEx
             newExtDirs[extDirs.length] = new File(serverEnvironment.getServerBaseDir(), "lib/ext");
             serviceTarget.addService(org.jboss.as.server.deployment.Services.JBOSS_DEPLOYMENT_EXTENSION_INDEX,
                     new ExtensionIndexService(newExtDirs)).setInitialMode(ServiceController.Mode.ON_DEMAND).install();
-            final Boolean suspend = runningModeControl.getSuspend()!= null ? runningModeControl.getSuspend() : serverEnvironment.isStartSuspended();
+            final boolean suspend = runningModeControl.getSuspend()!= null ? runningModeControl.getSuspend() : serverEnvironment.isStartSuspended();
+            final boolean gracefulStartup = serverEnvironment.isStartGracefully();
+            assert !suspend || gracefulStartup;
```

Comment:
This was a bad thing for me to have written. It might be ok but only if we've validated the user input elsewhere so this is basically asserting that that elsewhere did what it wanted. And TBH I don't think that 'validation elsewhere' is readily doable; this code is probably the place to validate and decide what to do.

What to do == TBD!

Reply:
Re the TBD, we've been chatting and I think if 'suspend=true' and 'gracefulStartup=false' then 'suspend' should win, but with logging.

The reason is because gracefulStartup can only be set from the cmd line, while the start-mode can be changed during a reload. So there can be workflows (odd ones, but they exist) where the user wants the long term setting to be graceless but temporarily they want the server suspended. The basic use case for graceless is a multiple deployment scenario where app B needs to invoke on app A during start. So the odd workflows would involve cases where app B isn't deployed yet or has been undeployed, and the admin is doing stuff and reloading.